docs/codeql/codeql-language-guides/customizing-library-models-for-java.rst

```diff
@@ -0,0 +1,439 @@
+         - ["java.util.stream", "Stream", True, "map", "(Function)", "", "Argument[0].ReturnValue", "ReturnValue.Element", "value", "manual"]
+
+Reasoning:
+
+Since we are adding flow through a method, we need to add tuples to the **summaryModel** extensible predicate.
+Each tuple defines part of the flow that comprises the total flow through the **map** method.
+The first five values identify the callable (in this case a method) to be modeled as a summary.
+These are the same for both of the rows above as we are adding two summaries for the same method.
+
+- The first value **java.util.stream** is the package name.
+- The second value **Stream** is the class (type) name.
+- The third value **True** is a flag that indicates whether or not the summary also applies to all overrides of the method.
+- The fourth value **map** is the method name.
+- The fifth value **Function** is the method input type signature.
+
+The sixth value is only relevant internally and can be omitted in most use cases.
+The remaining values are used to define the **access path**, the **kind**, and the **provenance** (origin) of the summary definition.
+
+- The seventh value is the access path to the **input** (where data flows from).
+- The eighth value is the access path to the **output** (where data flows to).
+
+For the first row the
+
+- The seventh value is **Argument[this].Element**, which is the access path to the elements of the qualifier (the elements of the stream **s** in the example).
+- The eight value is **Argument[0].Parameter[0]**, which is the access path to the first parameter of the **Function** argument of **map** (the lambda parameter **e** in the example).
+
+For the second row the
+
+- The seventh value is **Argument[0].ReturnValue**, which is the access path to the return value of the **Function** argument of **map** (the return value of the lambda in the example).
+- The eighth value is **ReturnValue.Element**, which is the access path to the elements of the return value of **map** (the elements of the stream **l** in the example).
+
+The remaining values for both rows
+
+- The ninth value **value** is the kind of the flow. **value** means that the value is preserved.
+- The tenth value **manual** is the provenance of the summary, which is used to identify the origin of the summary.
+
+That is, the first row models that there is value flow from the elements of the qualifier stream into the first argument of the Function provided to **map** and the second row models that there is value flow from the return value of the Function to the elements of the stream returned from **map**.
+
+Example: Add a **neutral** method.
+----------------------------------
+In this example we will show how to model the **now** method as being neutral.
+This is purely for completeness and has no impact on the analysis.
+A neutral model is used to define that there is no flow through a method.
+Please note that the neutral model for the **now** method is already added.
```

Comment:
I know this was already picked up in the resolved comment but it might be worth being explicit where it's already added, even if it's not consistent with the other examples by design.

Reply:
It turns out that it actually make sense now to write that it is added to the CodeQL java analysis as recent change incorporates the neutrals in the dataflow analysis (at least in terms of call targets). I will add the previous suggestion.